### PR TITLE
#4042 Pharma: extend customer import 

### PR DESCRIPTION
**Pull request by cristinamghita**
#4042 Pharma: extend customer import

### CODE REVIEW

File: de.metas.business/src/main/java/org/adempiere/impexp/bpartner/BPartnerImportProcess.java [hunk 1]
```diff
@@ -245,7 +246,11 @@ private final void createBPPrintFormatIfNeeded(@NonNull final I_I_BPartner impor
 					.adOrgId(importRecord.getAD_Org_ID())
 					.build());
 
-			final int AD_PrintFormat_ID =  Services.get(ISysConfigBL.class).getIntValue(BPARTNER_IMPORTPROCESS_BPPrintFormat, 1000015);
+			final int AD_PrintFormat_ID =  Services.get(ISysConfigBL.class).getIntValue(BPARTNER_IMPORTPROCESS_BPPrintFormatId, -1);
+			if (AD_PrintFormat_ID <= 0 )
+			{
+				throw new AdempiereException(BPARTNER_IMPORTPROCESS_BPPrintFormatId_ErrorMsg);
```
Comment (teosarca) [inline de.metas.business/src/main/java/org/adempiere/impexp/bpartner/BPartnerImportProcess.java]: your message won't be translated.
you need to:
* use `@`, before and after
* or use org.adempiere.exceptions.AdempiereException.AdempiereException(String, Object[])
* or use org.adempiere.exceptions.AdempiereException.AdempiereException(ITranslatableString) and de.metas.i18n.IMsgBL.getTranslatableMsgText(String, Object...)

Comment (cristinamghita) [reply de.metas.business/src/main/java/org/adempiere/impexp/bpartner/BPartnerImportProcess.java]: Thx! Fixed! I totally forgot about @@

Comment (teosarca) [reply de.metas.business/src/main/java/org/adempiere/impexp/bpartner/BPartnerImportProcess.java]: or better use AdempiereException.ofADMessage (I just introduced it in latest master)

File: de.metas.business/src/main/java/org/adempiere/impexp/bpartner/BPCreditLimitImportHelper.java [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+		return new BPCreditLimitImportHelper();
+	}
+
+	private BPCreditLimitImportHelper()
+	{
+	}
+
+	public final void importRecord(final I_I_BPartner importRecord)
+	{
+		I_C_BPartner_CreditLimit bpCreditLimit = null;
```
Comment (metas-ts) [inline de.metas.business/src/main/java/org/adempiere/impexp/bpartner/BPCreditLimitImportHelper.java]: don't we need to handle that case that there is already a credit limit that just needs to be updated?

File: de.metas.business/src/main/java/org/adempiere/impexp/bpartner/BPartnerImportTableSqlUpdater.java [hunk 3]
```diff
@@ -196,6 +202,27 @@ private void dbUpdateGreetings(final String whereClause)
 		logger.info("Invalid Greeting={}", no);
 	}
 
+	private void dbUpdateJobs(final String whereClause)
+	{
+		StringBuilder sql;
+		int no;
+		sql = new StringBuilder("UPDATE I_BPartner i "
```
Comment (metas-ts) [inline de.metas.business/src/main/java/org/adempiere/impexp/bpartner/BPartnerImportTableSqlUpdater.java]: is there a unique constraint in `C_Job` that makes sure we can have just one matching C_Job record?

File: de.metas.business/src/main/java/org/adempiere/impexp/bpartner/BPartnerImportTableSqlUpdater.java [hunk 4]
```diff
@@ -305,6 +332,54 @@ private void dbUpdatePO_PaymentTerms(final String whereClause)
 		logger.info("Invalid PO_PaymentTerm={}", no);
 	}
 
+	private void dbUpdateC_Aggregtions(final String whereClause)
+	{
+		StringBuilder sql;
+		int no;
+		sql = new StringBuilder("UPDATE I_BPartner i "
+				+ "SET C_Aggregation_ID=(SELECT C_Aggregation_ID FROM C_Aggregation a"
```
Comment (metas-ts) [inline de.metas.business/src/main/java/org/adempiere/impexp/bpartner/BPartnerImportTableSqlUpdater.java]: same question as for `C_Job` (and i think there is none)

File: de.metas.vertical.pharma/src/main/java/de/metas/impexp/bpartner/PharmaImportPartnerInterceptor.java [hunk 5]
```diff
@@ -51,19 +53,82 @@ private PharmaImportPartnerInterceptor()
...
 		}
+		//
+		if (!Check.isEmpty(ibpartner.getRegion(),true))
+		{
+			bpartner.setRegion(ibpartner.getRegion());
+		}
+		//
+		if (!Check.isEmpty(ibpartner.getSalesResponsible(),true))
+		{
+			bpartner.setPharmaproductpermlaw52(ibpartner.getPharmaproductpermlaw52());
```
Comment (metas-ts) [inline de.metas.vertical.pharma/src/main/java/de/metas/impexp/bpartner/PharmaImportPartnerInterceptor.java]: is this if with `SalesResponsible` and set with `Pharmaproductpermlaw52` correct?

File: de.metas.business/src/main/java/org/adempiere/impexp/bpartner/BPCreditLimitImportHelper.java (hunk 2 above)
Comment (cristinamghita) [reply de.metas.business/src/main/java/org/adempiere/impexp/bpartner/BPCreditLimitImportHelper.java]: Yes, you are right!

File: de.metas.vertical.pharma/src/main/java/de/metas/impexp/bpartner/PharmaImportPartnerInterceptor.java (hunk 5 above)
Comment (cristinamghita) [reply de.metas.vertical.pharma/src/main/java/de/metas/impexp/bpartner/PharmaImportPartnerInterceptor.java]: no! OMG, how did you see it?!

File: de.metas.business/src/main/java/org/adempiere/impexp/bpartner/BPartnerImportTableSqlUpdater.java (hunk 3 above)
Comment (cristinamghita) [reply de.metas.business/src/main/java/org/adempiere/impexp/bpartner/BPartnerImportTableSqlUpdater.java]: I think that should not happen, but is possible, so we should prevent.

File: de.metas.vertical.pharma/src/main/java/de/metas/impexp/bpartner/PharmaImportPartnerInterceptor.java (hunk 5 above)
Comment (metas-ts) [reply de.metas.vertical.pharma/src/main/java/de/metas/impexp/bpartner/PharmaImportPartnerInterceptor.java]: luck, i guess :-)

File: de.metas.adempiere.adempiere/base/src/main/java/org/adempiere/bpartner/service/BPartnerCreditLimitRepository.java [hunk 6]
```diff
@@ -118,4 +119,18 @@ private CreditLimitType retrieveCreditLimitTypePOJO(final int C_CreditLimit_Type
...
+		return Services.get(IQueryBL.class)
+				.createQueryBuilder(I_C_BPartner_CreditLimit.class)
+				.addEqualsFilter(I_C_BPartner_CreditLimit.COLUMNNAME_C_BPartner_ID, bpartnerId)
+				.addEqualsFilter(I_C_BPartner_CreditLimit.COLUMNNAME_Processed, true)
+				.addEqualsFilter(I_C_BPartner_CreditLimit.COLUMNNAME_C_CreditLimit_Type_ID, typeId)
+				.addOnlyActiveRecordsFilter()
+				.addOnlyContextClient()
+				.create()
+				.stream()
+				.findFirst();
```
Comment (metas-ts) [inline de.metas.adempiere.adempiere/base/src/main/java/org/adempiere/bpartner/service/BPartnerCreditLimitRepository.java]: can there be multiple matching C_BPartner_CreditLimits?
if yes, please make sure (by ordering) that we get the right one
if no, please add a UC if non exists yet, and use findFirstOnly

File: de.metas.business/src/main/java/org/adempiere/impexp/bpartner/BPartnerImportTableSqlUpdater.java (hunk 3 above)
Comment (metas-ts) [reply de.metas.business/src/main/java/org/adempiere/impexp/bpartner/BPartnerImportTableSqlUpdater.java]: thx, but what i had im mind was not to get *some* `C_Job` but rather add a UC to the C_Job table.
Maybe you can add one on `(Name, AD_Org_ID)` and then also filter for I_BPartner.AD_Org_ID if that makes sense

File: de.metas.business/src/main/java/org/adempiere/impexp/bpartner/BPartnerImportTableSqlUpdater.java (hunk 4 above)
Comment (metas-ts) [reply de.metas.business/src/main/java/org/adempiere/impexp/bpartner/BPartnerImportTableSqlUpdater.java]: again, same as with `C_Job`

File: de.metas.adempiere.adempiere/base/src/main/java/org/adempiere/bpartner/service/BPartnerCreditLimitRepository.java (hunk 6 above)
Comment (cristinamghita) [reply de.metas.adempiere.adempiere/base/src/main/java/org/adempiere/bpartner/service/BPartnerCreditLimitRepository.java]: Yes, can be

**Comment by cristinamghita** [comment]
@metas-ts done!

**Comment by teosarca** [comment]
@cristinamghita  pls merge it from master. your build is not stable.

File: de.metas.adempiere.adempiere/migration/src/main/sql/postgresql/system/10-de.metas.adempiere/5494240_sys_gh4402_AddUniqueIndexForC_JobAndC_Aggregation.sql [hunk 7]
```diff
@@ -0,0 +1,2 @@
+CREATE UNIQUE INDEX JobName ON C_Job(AD_Org_ID, Name);
```
[the remark this answers is not in the document]
Comment (cristinamghita) [reply de.metas.adempiere.adempiere/migration/src/main/sql/postgresql/system/10-de.metas.adempiere/5494240_sys_gh4402_AddUniqueIndexForC_JobAndC_Aggregation.sql]: Done!

**Pull request merged**